src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdCaseUpdater.java

```diff
@@ -0,0 +1,228 @@
+    private final ServiceResponseParser serviceResponseParser;
+    private final AuthTokenGenerator s2sTokenGenerator;
+    private final CoreCaseDataApi coreCaseDataApi;
+    private final ExceptionRecordFinalizer exceptionRecordFinalizer;
+
+    public CcdCaseUpdater(
+        CaseUpdateClient caseUpdateClient,
+        ServiceResponseParser serviceResponseParser,
+        AuthTokenGenerator s2sTokenGenerator,
+        CoreCaseDataApi coreCaseDataApi,
+        ExceptionRecordFinalizer exceptionRecordFinalizer
+    ) {
+        this.caseUpdateClient = caseUpdateClient;
+        this.serviceResponseParser = serviceResponseParser;
+        this.s2sTokenGenerator = s2sTokenGenerator;
+        this.coreCaseDataApi = coreCaseDataApi;
+        this.exceptionRecordFinalizer = exceptionRecordFinalizer;
+    }
+
+    @SuppressWarnings({"squid:S2139", "unchecked"}) // squid for exception handle + logging
+    public ProcessResult updateCase(
+        ExceptionRecord exceptionRecord,
+        ServiceConfigItem configItem,
+        boolean ignoreWarnings,
+        String idamToken,
+        String userId,
+        CaseDetails existingCase
+    ) {
+        log.info(
+            "Start updating case for service {} with case Id {} from exception record {}",
+            configItem.getService(),
+            existingCase.getId(),
+            exceptionRecord.id
+        );
+
+        try {
+            String s2sToken = s2sTokenGenerator.generate();
+
+            SuccessfulUpdateResponse updateResponse = caseUpdateClient.updateCase(
+                configItem.getUpdateUrl(),
+                existingCase,
+                exceptionRecord,
+                s2sToken
+            );
+
+            log.info(
+                "Successfully called case update endpoint of service {} to update case with case Id {} "
+                    + "based on exception record ref {}",
+                configItem.getService(),
+                existingCase.getId(),
+                exceptionRecord.id
+            );
+
+            if (!ignoreWarnings && !updateResponse.warnings.isEmpty()) {
+                // do not log warnings
+                return new ProcessResult(updateResponse.warnings, emptyList());
+            } else {
+                updateCaseInCcd(
+                    configItem.getService(),
+                    ignoreWarnings,
+                    idamToken,
+                    s2sToken,
+                    userId,
+                    exceptionRecord,
+                    updateResponse.caseDetails,
+                    existingCase
+                );
+
+                return new ProcessResult(
+                    exceptionRecordFinalizer.finalizeExceptionRecord(existingCase.getData(), existingCase.getId())
+                );
+            }
+        } catch (BadRequest exception) {
+            throw new CallbackException(
+                format(
+                    "Failed to call %s service Case Update API to update case with case Id %s "
```

Comment:
This `catch` is for the try that wraps two calls, but we're assuming the exception concerns the first one. I think we should catch this exception closer to calling update endpoint (it doesn't have to be limited to `BadRequest` - can be the top-most exception for error response, because we're catching it to log a failure), and also handle `UnprocessableEntity` in that place instead of here.

Reply:
I  have played a bit with different other options - unfortunately this try - catch layout looks to be the best out of other not ideal layouts.

Reply:
It's not just about the layout @apereverzin.
`coreCaseDataApi.startEventForCaseWorker(...)` from line 73 can also throw `BadRequest` and then the message in this catch will be incorrect.

Reply:
Have fixed